File: 21553 - Sai Nandan D/Week 8/986_Interval List Intersections.py
class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:        
        i, j = 0, 0
        result = []        
        intersection = lambda f, s: f[0] <= s[1] and s[0] <= f[1]        
        while i < len(firstList) and j < len(secondList):
            f, s = firstList[i], secondList[j]
            if intersection(f, s):
                result.append([max(f[0], s[0]), min(f[1], s[1])])
            if f[1] <= s[1]:
                i += 1
            else:
                j += 1                
        return result

File: 21561 - Suvam Tamang/Week 8/986_Interval List Intersections.java
class Solution {
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {
        if(firstList.length == 0 || secondList.length == 0)
            return new int[][]{};
        List<int[]> res = new ArrayList<>();
        int i = 0, j = 0;
        int leftIntersection, rightIntersection;
        while(i < firstList.length && j < secondList.length){
            leftIntersection = Math.max(firstList[i][0], secondList[j][0]);
            rightIntersection = Math.min(firstList[i][1], secondList[j][1]);
            
            if(rightIntersection >= leftIntersection)
                res.add(new int[]{leftIntersection, rightIntersection});
            
            if(firstList[i][1] == rightIntersection) i++;
            if(secondList[j][1] == rightIntersection) j++;
        }
        
        return res.toArray(new int[res.size()][2]);
    }
}
